File: solutions/0124_Binary_Tree_Maximum_Path_Sum/dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.ret = -1<<32
        self.dfs(root)
        return self.ret

    def dfs(self, root):
        if not root:
            return 0
        l = self.dfs(root.left)
        r = self.dfs(root.right)
        cur_sum = root.val + max(0, l) + max(0, r)
        self.ret = max(self.ret, cur_sum)
        return max(0, root.val + max(0, l, r))

File: solutions/0337_House_Robber_III/dfs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        Pair<Integer, Integer> ret = dfs(root);
        return Math.max(ret.getKey(), ret.getValue());
    }

    // pick, not pick
    private Pair<Integer, Integer> dfs(TreeNode node) {
        if (node == null) return new Pair<>(0, 0);
        Pair<Integer, Integer> p1 = dfs(node.left);
        Pair<Integer, Integer> p2 = dfs(node.right);
        return new Pair<>(p1.getValue() + p2.getValue() + node.val,
                          Math.max(p1.getKey(), p1.getValue()) +
                          Math.max(p2.getKey(), p2.getValue()));
    }
}
